src/test/java/com/marklogic/spark/reader/file/ReadMlcpArchiveFilesTest.java

```diff
@@ -1,43 +1,169 @@
 package com.marklogic.spark.reader.file;
 
+import com.marklogic.junit5.XmlNode;
 import com.marklogic.spark.AbstractIntegrationTest;
+import com.marklogic.spark.ConnectorException;
 import com.marklogic.spark.Options;
 import org.apache.spark.sql.Dataset;
 import org.apache.spark.sql.Row;
 import org.junit.jupiter.api.Test;
+import scala.collection.mutable.WrappedArray;
 
-import java.util.HashSet;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 class ReadMlcpArchiveFilesTest extends AbstractIntegrationTest {
 
+    /**
+     * Depends on an MLCP archive file containing the two XML documents and their metadata entries as produced by
+     * TestUtil.insertTwoDocumentsWithAllMetadata.
+     */
     @Test
     void readMlcpArchiveFile() {
-        Dataset<Row> reader = newSparkSession().read()
+        List<Row> rows = newSparkSession().read()
             .format(CONNECTOR_IDENTIFIER)
             .option(Options.READ_FILES_TYPE, "mlcp_archive")
-            .load("src/test/resources/mlcp-archive-files/sample-XML.zip");
-        verifyFileRows(reader.collectAsList());
-    }
-
-    private void verifyFileRows(List<Row> rows) {
-        HashSet<String> hashSet = new HashSet<>();
-        hashSet.add("/alexander");
-        hashSet.add("/alexander.metadata");
-        hashSet.add("/betsy");
-        hashSet.add("/betsy.metadata");
-        hashSet.add("/george");
-        hashSet.add("/george.metadata");
-        hashSet.add("/martha");
-        hashSet.add("/martha.metadata");
-        hashSet.add("/mary");
-        hashSet.add("/mary.metadata");
-
-        for(int i =0; i<10; i++){
-            String temp = rows.get(i).getString(0).substring(rows.get(i).getString(0).lastIndexOf("/"));
-            assert(hashSet.contains(temp));
-            hashSet.remove(temp);
-        }
-        assert(hashSet.isEmpty());
+            .load("src/test/resources/mlcp-archive-files/files-with-all-metadata.mlcp.zip")
+            .collectAsList();
+
+        assertEquals(2, rows.size(), "The .metadata entries should not be returned as rows, but rather should be " +
+            "used to populate the metadata columns in the rows for the documents they're associated with.");
+
+        verifyFirstRow(rows.get(0));
+        verifySecondRow(rows.get(1));
+    }
+
+    @Test
+    void mlcpArchivesContainingAllFourTypesOfDocuments() {
+        List<Row> rows = newSparkSession().read()
+            .format(CONNECTOR_IDENTIFIER)
+            .option(Options.READ_FILES_TYPE, "mlcp_archive")
+            .load("src/test/resources/mlcp-archive-files/all-four-document-types")
+            .collectAsList();
+
+        assertEquals(4, rows.size());
+
+        assertEquals("JSON", rows.get(0).getString(2));
+        assertEquals("XML", rows.get(1).getString(2));
```

Comment:
Looks like the test failed here.
Is collectAsList ordered?

Reply:
Interesting - it should be since it's a single partition reader, but maybe Spark is still free to toss those rows in the dataset in any order? I'll add a "sort" to it for the sole purpose of making the assertions reliable.